Add ability to settle a NativePromise with a function which will be run on the dispatcher given to then/whenSettled

jyavenard opened the pull request:
#### e323dca2b77f79cb09bfc9f457f721deac22c845
<pre>
Add ability to settle a NativePromise with a function which will be run on the dispatcher given to then/whenSettled
<a href="https://bugs.webkit.org/show_bug.cgi?id=270282">https://bugs.webkit.org/show_bug.cgi?id=270282</a>
<a href="https://rdar.apple.com/123830373">rdar://123830373</a>

Reviewed by Youenn Fablet.

We add a settleWithFunction method to NativePromiseProducer.
For convenience, you can also pass the Function to settle().
The function provided will be run for in listener&apos;s dispatcher
and the result be then passed to the callback.

Fly-by: Make accessing the promise&apos;s result more explicit (such as when is the result accessed by reference or being moved)

Added API tests.

* Source/WTF/wtf/NativePromise.h:
* Tools/TestWebKitAPI/Tests/WTF/NativePromise.cpp:
(TestWebKitAPI::TEST):

Canonical link: <a href="https://commits.webkit.org/275720@main">https://commits.webkit.org/275720@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/cb7f03778ea364bccbe8a479b10712289c646088

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [⏳ 🧪 style ](https://ews-build.webkit.org/#/builders/Style-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 ios ](https://ews-build.webkit.org/#/builders/iOS-17-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 mac ](https://ews-build.webkit.org/#/builders/macOS-Monterey-Release-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 wpe ](https://ews-build.webkit.org/#/builders/WPE-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 wincairo ](https://ews-build.webkit.org/#/builders/WinCairo-EWS "Waiting in queue, processing has not started yet") 
| [⏳ 🧪 bindings ](https://ews-build.webkit.org/#/builders/Bindings-Tests-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 ios-sim ](https://ews-build.webkit.org/#/builders/iOS-17-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 mac-AS-debug ](https://ews-build.webkit.org/#/builders/macOS-AppleSilicon-Sonoma-Debug-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 wpe-wk2 ](https://ews-build.webkit.org/#/builders/WPE-Build-EWS "Waiting in queue, processing has not started yet") | 
| [⏳ 🧪 webkitperl ](https://ews-build.webkit.org/#/builders/WebKitPerl-Tests-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 ios-wk2 ](https://ews-build.webkit.org/#/builders/iOS-17-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 api-mac ](https://ews-build.webkit.org/#/builders/macOS-Monterey-Release-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 api-wpe ](https://ews-build.webkit.org/#/builders/WPE-Build-EWS "Waiting in queue, processing has not started yet") | 
| [⏳ 🧪 webkitpy ](https://ews-build.webkit.org/#/builders/WebKitPy-Tests-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 ios-wk2-wpt ](https://ews-build.webkit.org/#/builders/iOS-17-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 mac-wk1 ](https://ews-build.webkit.org/#/builders/macOS-Monterey-Release-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 wpe-skia ](https://ews-build.webkit.org/#/builders/WPE-Skia-Build-EWS "Waiting in queue, processing has not started yet") | 
| [⏳ 🛠 🧪 jsc ](https://ews-build.webkit.org/#/builders/JSC-Tests-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 api-ios ](https://ews-build.webkit.org/#/builders/iOS-17-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 mac-wk2 ](https://ews-build.webkit.org/#/builders/macOS-Monterey-Release-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 gtk ](https://ews-build.webkit.org/#/builders/GTK-Build-EWS "Waiting in queue, processing has not started yet") | 
| [⏳ 🛠 🧪 jsc-arm64 ](https://ews-build.webkit.org/#/builders/JSC-Tests-arm64-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 tv ](https://ews-build.webkit.org/#/builders/tvOS-17-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 mac-AS-debug-wk2 ](https://ews-build.webkit.org/#/builders/macOS-AppleSilicon-Sonoma-Debug-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 gtk-wk2 ](https://ews-build.webkit.org/#/builders/GTK-Build-EWS "Waiting in queue, processing has not started yet") | 
| [⏳ 🧪 services ](https://ews-build.webkit.org/#/builders/Services-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 tv-sim ](https://ews-build.webkit.org/#/builders/tvOS-17-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 mac-wk2-stress ](https://ews-build.webkit.org/#/builders/macOS-Monterey-Release-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 api-gtk ](https://ews-build.webkit.org/#/builders/GTK-Build-EWS "Waiting in queue, processing has not started yet") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9526 "Built successfully and passed tests") | [⏳ 🛠 watch ](https://ews-build.webkit.org/#/builders/watchOS-10-Build-EWS "Waiting in queue, processing has not started yet") | | [⏳ 🛠 jsc-armv7 ](https://ews-build.webkit.org/#/builders/JSC-ARMv7-32bits-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [⏳ 🛠 watch-sim ](https://ews-build.webkit.org/#/builders/watchOS-10-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | | [⏳ 🧪 jsc-armv7-tests ](https://ews-build.webkit.org/#/builders/JSC-ARMv7-32bits-Build-EWS "Waiting in queue, processing has not started yet") | 
<!--EWS-Status-Bubble-End-->

squelart commented on Source/WTF/wtf/NativePromise.h:
```diff
@@ -1141,10 +1160,25 @@ class NativePromise final : public NativePromiseBase, public ConvertibleToNative
+    {
+        // Only called by SettleFunction on the target's queue once all operations are complete and settled.
+        // So we don't really need to hold the lock to access the value.
+        Locker lock { m_lock };
```
Is this `m_lock` the same as the one that's `DropLockForScope`'d in the caller above?
If yes, to avoid the extra lockings (especially if they're not really necessary), could you instead call `hasRunnable()` and `resultRunnable()` before the `DropLockForScope`.
And maybe even combine them into one function that return a nullable pointer?

jyavenard replied:
The lock really does nothing, we return a reference here too.

jyavenard replied:
also, we do need a lock otherwise it will not compile due to the `WTF_GUARDED_BY_LOCK`

jyavenard replied:
I've tweaked things a bit so that if a copy/move is to occur, it happens under the lock.

squelart reviewed (commented):
Unofficial r+ from me, with one and a half nits for your consideration.

youennf commented on Source/WTF/wtf/NativePromise.h:
```diff
@@ -713,25 +721,35 @@ class NativePromise final : public NativePromiseBase, public ConvertibleToNative
-                    processResult(promise.result());
+                    if (promise.hasRunnable()) {
+                        ASSERT(IsExclusive);
+                        auto resultRunnable = WTFMove(promise.resultRunnable());
```
This WTFMove is a bit odd.
Should we encapsulate this in the `result()` call instead that would call resultRunnable() and store the returned value in m_result?
Or have a takeResultRunnable() maybe.

jyavenard replied:
to do so would require a more intensive change to the code rather than re-use what's already there as much as possible

jyavenard replied:
The runnable is already a detail of Storage.

There are limitations to how C++ can deal between returned value types; this isn't a templated method: you also can't template and have automatic deduction where the only difference in the code is the return value.

jyavenard replied:
As discussed off-line; I have made the move semantics a bit clearer , with the use of a takeResult/result and takeResultRunnable

youennf commented on Source/WTF/wtf/NativePromise.h:
```diff
@@ -1192,42 +1229,54 @@ class NativePromise final : public NativePromiseBase, public ConvertibleToNative
-                return m_result.has_value();
-            else
-                return m_result->result.has_value();
+            return std::holds_alternative<Result>(optionalResult());
```
has_value() is a bit odd now since it does not account for the possiblity to get a result from the runnable.
This has an impact on `isResolved()` getter for instance.
I would think that a promise that has a runnable actually has a value.

jyavenard replied:
I have renamed the method to `hasResult()`

youennf reviewed (commented):
I would be tempted to keep the runnable as a detail of Storage if possible.
We could lock, call the settle function, get the result and store it.
If we ever want to support calling the settle function more than once, we can always add a separate method for that in the future.

jyavenard committed:
Add ability to settle a NativePromise with a function which will be run on the dispatcher given to then/whenSettled
https://bugs.webkit.org/show_bug.cgi?id=270282
rdar://123830373

Reviewed by Youenn Fablet.

We add a settleWithFunction method to NativePromiseProducer.
For convenience, you can also pass the Function to settle().
The function provided will be run for in listener's dispatcher
and the result be then passed to the callback.

Fly-by: Make accessing the promise's result more explicit (such as when is the result accessed by reference or being moved)

Added API tests.

* Source/WTF/wtf/NativePromise.h:
* Tools/TestWebKitAPI/Tests/WTF/NativePromise.cpp:
(TestWebKitAPI::TEST):

Canonical link: https://commits.webkit.org/275720@main
